{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:41.998781Z",
     "start_time": "2020-09-15T20:32:39.797387Z"
    }
   },
   "outputs": [],
   "source": [
    "#export\n",
    "import json\n",
    "import sys\n",
    "import os\n",
    "import os.path\n",
    "from pathlib import Path\n",
    "from subprocess import call\n",
    "import subprocess\n",
    "import sys # system libraries, like arguments (argv)\n",
    "import re # regular expressions\n",
    "import pandas as pd\n",
    "import glob\n",
    "import os.path\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "import scipy.stats as stats\n",
    "# import researchpy as rp\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "import statsmodels\n",
    "# print(statsmodels.__version__)  \n",
    "# 0.8.0rc1\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import fdrcorrection0\n",
    "# Pulled from https://pythonfordatascience.org/anova-python/\n",
    "\n",
    "\n",
    "\n",
    "from timeit import default_timer as timer\n",
    "import time\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:42.005388Z",
     "start_time": "2020-09-15T20:32:42.001619Z"
    }
   },
   "outputs": [],
   "source": [
    "#export\n",
    "def removeFile(filename):\n",
    "    try:\n",
    "        os.remove(filename)\n",
    "    except OSError:\n",
    "        pass   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:42.012008Z",
     "start_time": "2020-09-15T20:32:42.008238Z"
    }
   },
   "outputs": [],
   "source": [
    "#export\n",
    "def makeFolders(folder_list):\n",
    "    for directory in folder_list:\n",
    "        if not directory: # Make sure not an empty string \"\"\n",
    "            continue\n",
    "        else:\n",
    "            if not os.path.exists(directory):\n",
    "                os.makedirs(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:43.563998Z",
     "start_time": "2020-09-15T20:32:42.014392Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rp' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-448dc59de03d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m# Gettin summary statistics\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0mrp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msummary_cont\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'libido'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rp' is not defined"
     ]
    }
   ],
   "source": [
    "# Anova test\n",
    "\n",
    "\n",
    "\n",
    "# Loading data\n",
    "df = pd.read_csv(\"https://raw.githubusercontent.com/Opensourcefordatascience/Data-sets/master/difficile.csv\")\n",
    "df.drop('person', axis= 1, inplace= True)\n",
    "\n",
    "# Recoding value from numeric to string\n",
    "df['dose'].replace({1: 'placebo', 2: 'low', 3: 'high'}, inplace= True)\n",
    "    \n",
    "# Gettin summary statistics\n",
    "rp.summary_cont(df['libido'])\n",
    "\n",
    "print(df.head())\n",
    "print(df.dtypes)\n",
    "print(df.shape)\n",
    "\n",
    "results = ols('libido ~ C(dose)', data=df).fit()\n",
    "results.summary()\n",
    "print(results.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:43.565661Z",
     "start_time": "2020-09-15T20:32:42.453Z"
    }
   },
   "outputs": [],
   "source": [
    "class FilterANOVA():\n",
    "  \n",
    "    def __init__(self, \n",
    "         NF_out, \n",
    "         NF_path,\n",
    "         col_data, \n",
    "         title,\n",
    "         p_adj,\n",
    "         meta_comparison=\"condition\",\n",
    "         searchlist = None,\n",
    "         binRank = None,\n",
    "         min_length = 100):\n",
    "        \n",
    "        \n",
    "        self.NF_path = NF_path\n",
    "        self.p_adj = p_adj\n",
    "        self.searchlist = searchlist\n",
    "        self.binRank = binRank\n",
    "        self.min_length = min_length\n",
    "        self.meta_comparison = meta_comparison\n",
    "        \n",
    "        self.title = title\n",
    "        self.NF_out = NF_out\n",
    "        self.wkdir = f'{self.NF_out}/unmapped/final'\n",
    "        self.col_data = col_data\n",
    "        self.df_col = pd.read_csv(self.col_data,sep=\"\\t\",names = [\"sample\",\"condition\"])\n",
    "        self.condition_list = self.df_col[\"condition\"].unique() \n",
    "        \n",
    "        \n",
    "  \n",
    "    def searchDF(self, df, HAS, rank, name):\n",
    "        df = df.fillna(\"NA\")\n",
    "        if HAS:\n",
    "            df = df.loc[df[rank]==name]\n",
    "        else:\n",
    "            df = df.loc[df[rank]!=name]\n",
    "        return df\n",
    "        \n",
    "        \n",
    "    def Filter(self, df):\n",
    "        if self.searchlist == None or self.searchlist == \"contig\":\n",
    "            return df\n",
    "        else:\n",
    "            for search in self.searchlist:           \n",
    "                if len(df) > 0:   \n",
    "                    HAS = search[0]\n",
    "                    rank = search[1]\n",
    "                    name = search[2]\n",
    "                    \n",
    "                    print(f\"Searching with {HAS}:{rank}:{name}\")\n",
    "                    df = self.searchDF(df=df, HAS=HAS, rank=rank,name=name) \n",
    "            return df\n",
    "        \n",
    "        \n",
    "     \n",
    "    def runANOVA(self,df):\n",
    "        print(f\"Running Ttest with {len(df)}\")\n",
    "                \n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "    def runSearch(self):\n",
    "        files = glob.glob(f'{self.wkdir}/pileupCoverageNormalizedMatched/*.txt')\n",
    "        meta_comparison = self.meta_comparison\n",
    "\n",
    "        out = f'{self.NF_out}/FINAL_OUT/{self.title}'\n",
    "        minout = f'{out}/minlength_{self.min_length}'\n",
    "        makeFolders([f'{self.NF_out}/FINAL_OUT/',out])\n",
    "\n",
    "        df = pd.read_csv(files[0],sep=\"\\t\",low_memory=False)\n",
    "        df = self.Filter(df) \n",
    "        df = df[df[binRank] != \"NA\"]\n",
    "        df = df[df[binRank].notna()]\n",
    "        df_col = self.df_col\n",
    "                \n",
    "\n",
    "        \n",
    "        samples = df_col[\"sample\"].to_list()\n",
    "        df_col = df_col.set_index(\"sample\")\n",
    "        meta_data = list(df_col)\n",
    "        \n",
    "        \n",
    "        #Change dataframe to long\n",
    "        df = df.T\n",
    "\n",
    "        df_samples = df.loc[ samples , : ]\n",
    "        df_binrank = df[df.index.str.match(self.binRank)]\n",
    "\n",
    "        df = pd.concat((df_binrank,df_samples))        \n",
    "        new_header = df.iloc[0] #grab the first row for the header\n",
    "        df = df[1:] #take the data less the header row\n",
    "        df.columns = new_header #set the header row as the df header\n",
    "        contigs = df.columns\n",
    "        num_contigs = len(list(contigs))\n",
    "        df = pd.merge(df,df_col, left_on=df.index, right_on=\"sample\")\n",
    "        df = df.set_index(\"sample\")\n",
    "        \n",
    "        \n",
    "        #Loop through and run ANOVA for each contig\n",
    "        \n",
    "        \n",
    "        x = 0\n",
    "        contig = contigs[x]\n",
    "        df2 = df[[contig]+[meta_comparison]].copy()\n",
    "        df2[contig] = df2[contig].astype(float)\n",
    "        print(df2.dtypes)      \n",
    "\n",
    "\n",
    "        df2[\"temp\"] = df2[contig]\n",
    "        results = ols(f'temp ~ C({meta_comparison})', data=df2).fit()\n",
    "#         print(results.summary()) \n",
    "\n",
    "        aov_table = sm.stats.anova_lm(results, typ=2)\n",
    "        print(aov_table)\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "#         # With scipy\n",
    "#         comparisons = list(df2[meta_comparison].unique())\n",
    "\n",
    "#         lc = []\n",
    "#         for comparison in comparisons:\n",
    "#             temp_df = df2[df2[meta_comparison]==comparison].copy()\n",
    "#             numbers = temp_df[contig]\n",
    "#             lc.append(numbers)\n",
    "\n",
    "#         fvalue, pvalue = stats.f_oneway(lc[0],lc[1],lc[2])\n",
    "#         print(fvalue,pvalue)\n",
    "        ########\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "#         print(df2.head(10))\n",
    "        df.to_csv(\"test.txt\",sep=\"\\t\")\n",
    "\n",
    "        \n",
    "        \n",
    "#         dfs = []\n",
    "#         for f in files:\n",
    "#             print(\"---------------\")\n",
    "#             file = os.path.basename(f)\n",
    "#             print(\"out:\", out)\n",
    "#             print(f' Working on file {file}')\n",
    "#             df = pd.read_csv(f,sep=\"\\t\",low_memory=False)\n",
    "#             print(f\"Starting contigs: {len(df)}\")\n",
    "\n",
    "#             df = self.Filter(df)\n",
    "#             print(df)\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:44.059700Z",
     "start_time": "2020-09-15T20:32:44.047411Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'FilterANOVA' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-63add780b824>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0mtitle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"OUT_All_Contigs_bin_species_ANOVA\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m F = FilterANOVA(NF_out=NF_out,\n\u001b[0m\u001b[1;32m     17\u001b[0m                 \u001b[0mNF_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNF_path\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m              \u001b[0mcol_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcol_data\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'FilterANOVA' is not defined"
     ]
    }
   ],
   "source": [
    "# NF = \"/Users/m/Google_Drive/Scripts/2019/biome/biome_shared/nextflow/RNAseq-Biome-Nextflow\"\n",
    "NF_path = \"/Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master\"\n",
    "\n",
    "NF_out = f\"{NF_path}/NF_OUT\" \n",
    "col_data = f\"{NF_path}/sample_table.txt\"\n",
    "df_col = pd.read_csv(col_data,sep=\"\\t\",names=[\"sample\",\"condition\"])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "searchlist = None\n",
    "binRank = \"contig\"\n",
    "\n",
    "title = \"OUT_All_Contigs_bin_species_ANOVA\"\n",
    "\n",
    "F = FilterANOVA(NF_out=NF_out,\n",
    "                NF_path = NF_path,\n",
    "             col_data = col_data, \n",
    "             title=title, \n",
    "             searchlist=searchlist,\n",
    "             meta_comparison = \"condition\",\n",
    "             binRank = binRank,\n",
    "             min_length = 100,\n",
    "             p_adj = 0.05\n",
    "            )\n",
    "\n",
    "F.runSearch()\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:44.423424Z",
     "start_time": "2020-09-15T20:32:44.393016Z"
    }
   },
   "outputs": [],
   "source": [
    "#export\n",
    "class FilterTtest():\n",
    "    \n",
    "    def __init__(self, \n",
    "         NF_out, \n",
    "         NF_path,\n",
    "         col_data, \n",
    "         title,\n",
    "         control,\n",
    "         treatment,\n",
    "         p_adj,\n",
    "         searchlist = None,\n",
    "         binRank = None,\n",
    "         min_length = 100):\n",
    "\n",
    "        self.NF_path = NF_path\n",
    "        self.p_adj = p_adj\n",
    "        self.searchlist = searchlist\n",
    "        self.binRank = binRank\n",
    "        self.control = control\n",
    "        self.treatment = treatment \n",
    "        self.min_length = min_length\n",
    "\n",
    "        self.title = title\n",
    "        self.NF_out = NF_out\n",
    "        self.wkdir = f'{self.NF_out}/unmapped/final'\n",
    "        self.col_data = col_data\n",
    "        self.df_col = pd.read_csv(self.col_data,sep=\"\\t\",names = [\"sample\",\"condition\"])\n",
    "        self.condition_list = self.df_col[\"condition\"].unique()\n",
    "\n",
    "        \n",
    "    def searchDF(self, df, HAS, rank, name):\n",
    "        df = df.fillna(\"NA\")\n",
    "        if HAS:\n",
    "            df = df.loc[df[rank]==name]\n",
    "        else:\n",
    "            \n",
    "            df = df.loc[df[rank]!=name]\n",
    "        return df\n",
    "        \n",
    "        \n",
    "    def Filter(self, df):\n",
    "        if self.searchlist == None:\n",
    "            return df\n",
    "        else:\n",
    "            for search in self.searchlist:           \n",
    "                if len(df) > 0:   \n",
    "                    HAS = search[0]\n",
    "                    rank = search[1]\n",
    "                    name = search[2]\n",
    "                    \n",
    "                    print(f\"Searching with {HAS}:{rank}:{name}\")\n",
    "                    df = self.searchDF(df=df, HAS=HAS, rank=rank,name=name) \n",
    "            return df\n",
    "\n",
    "        \n",
    "\n",
    "        \n",
    "    def runTtest(self,df):\n",
    "        print(f\"Running Ttest with {len(df)}\")\n",
    "        \n",
    "        # Get coverages by conditions \n",
    "        treatment_df = self.df_col[self.df_col[\"condition\"]==self.treatment]\n",
    "        treatment_list = list(treatment_df[\"sample\"])\n",
    "        control_df = self.df_col[self.df_col[\"condition\"]==self.control]\n",
    "        control_list = list(control_df[\"sample\"])\n",
    "        \n",
    "        df_treatment = df[treatment_list]\n",
    "        df_control = df[control_list]\n",
    "        \n",
    "        t_list = []\n",
    "        p_val = []\n",
    "        df[\"levine_statistic\"] = np.nan\n",
    "        df[\"levine_pval\"] = np.nan\n",
    "        df[\"t_statistic\"] = np.nan\n",
    "        df[\"t_pval\"] = np.nan\n",
    "#         df[\"t_padj\"] = 0.0\n",
    "        \n",
    "        df = df.copy()\n",
    "        for i in range(len(df)):\n",
    "            treatment_vals = df_treatment.iloc[i,:]\n",
    "            control_vals = df_control.iloc[i,:]\n",
    "            \n",
    "            lev = stats.levene(treatment_vals, control_vals)\n",
    "            df[\"levine_statistic\"].iloc[i] = lev[0]\n",
    "            df[\"levine_pval\"].iloc[i] = lev[1]\n",
    "        \n",
    "        \n",
    "            ttest = stats.ttest_ind(treatment_vals, control_vals)\n",
    "            df[\"t_statistic\"].iloc[i] = ttest[0]\n",
    "            df[\"t_pval\"].iloc[i] = ttest[1]\n",
    "        \n",
    "\n",
    "        #Benjamini hochberg fdr correction\n",
    "        df[\"t_padj\"] = np.nan\n",
    "        df_null = df[df[\"t_pval\"].isnull()]\n",
    "        df = df[~df[\"t_pval\"].isnull()] # Get everything not null because statsmodel cant handle it\n",
    "\n",
    "        padj = statsmodels.sandbox.stats.multicomp.fdrcorrection0(df[\"t_pval\"], alpha=0.05, method='indep', is_sorted=False)        \n",
    "        \n",
    "        df[\"t_padj_reject\"] = padj[0] \n",
    "        df[\"t_padj\"] = list(padj[1])\n",
    "        df = pd.concat([df,df_null], sort=False)\n",
    "\n",
    "#         print(df)\n",
    "\n",
    "    \n",
    "        cols = list(df)\n",
    "        \n",
    "        statcols = [\"contig\", \"name\", \"superkingdom\", \"kingdom\", \"phylum\", \"order\",\t\"family\", \"genus\", \"species\", \"blast_pident\", \"blast_sseqid\", \"blast_evalue\", \"taxid\", \"blast_stitle\", \"filename\", \"length\", \"fasta\"]\n",
    "        statcols = [\"levine_statistic\",\"levine_pval\",\"t_statistic\",\"t_pval\",\"t_padj\",\"t_padj_reject\"]\n",
    "        newcols = [x for x in cols if x not in statcols]\n",
    "        cols = statcols + newcols\n",
    "        df = df[cols]\n",
    "\n",
    "        \n",
    "        return df\n",
    "    \n",
    "    def runSearch(self):\n",
    "        files = glob.glob(f'{self.wkdir}/pileupCoverageNormalizedMatched/*.txt')\n",
    "\n",
    "        out = f'{self.NF_out}/FINAL_OUT/{self.title}'\n",
    "        ddout = f'{self.NF_out}/FINAL_OUT/{control}_{treatment}_doubledark'\n",
    "        minout = f'{out}/minlength_{self.min_length}'\n",
    "        makeFolders([f'{self.NF_out}/FINAL_OUT/',out, ddout])\n",
    "\n",
    "        empties = []\n",
    "        \n",
    "                    \n",
    "            #Run T test for no hits if it doesn't exist\n",
    "        if os.path.exists(f'{ddout}/{title}_dark_all_FG_NoHits.txt') == False:\n",
    "            df = pd.read_csv(f'{self.wkdir}/pileupCoverageNormalizedMatched/dark_all_FG_NoHits_pileupCoverageNormalizedMatched.tsv',sep=\"\\t\")\n",
    "            df = self.runTtest(df)\n",
    "            df_sig = df[df[\"t_padj_reject\"]==True]\n",
    "            f_sig = f'{ddout}/{control}_{treatment}_doubledark_double_dark_all_FG_NoHits_Significant.txt'\n",
    "            f_df = f'{ddout}/{control}_{treatment}_doubledark_double_dark_all_FG_NoHits.txt'\n",
    "            df_sig.to_csv(f_sig,sep=\"\\t\",index=None)\n",
    "            df.to_csv(f_df,sep=\"\\t\",index=None)            \n",
    "\n",
    "\n",
    "        \n",
    "        \n",
    "        dfs = []\n",
    "        for f in files:\n",
    "            print(\"---------------\")\n",
    "            file = os.path.basename(f)\n",
    "            print(\"out:\", out)\n",
    "            print(f' Working on file {file}')\n",
    "            df = pd.read_csv(f,sep=\"\\t\",low_memory=False)\n",
    "            print(f\"Starting contigs: {len(df)}\")\n",
    "            df = self.Filter(df)\n",
    "\n",
    "            \n",
    "            if len(df) > 0:\n",
    "                if self.binRank != None:\n",
    "                    df = df.groupby([self.binRank]).sum()\n",
    "                    print(f\"Binning by Rank {self.binRank}: Length: {len(df)}\")\n",
    "\n",
    "                \n",
    "                \n",
    "#                 print(df.head())\n",
    "                df = self.runTtest(df)\n",
    "#                 print(df.head())\n",
    "                df_sig = df[df[\"t_padj_reject\"]==True]\n",
    "                f_sig = f'{out}/{self.title}_Significant_{file}'\n",
    "                f_df = f'{out}/{self.title}_{file}'\n",
    "\n",
    "                self.df_sig = df_sig\n",
    "                if self.binRank != None:\n",
    "                    df_sig.to_csv(f_sig,sep=\"\\t\")\n",
    "                    df.to_csv(f_df,sep=\"\\t\")\n",
    "                else:\n",
    "                    df_sig.to_csv(f_sig,sep=\"\\t\",index=None)\n",
    "                    df.to_csv(f_df,sep=\"\\t\",index=None)\n",
    "                    \n",
    "                \n",
    "            else:\n",
    "                \n",
    "                empties.append(file)\n",
    "                \n",
    "#         if len(empties)  > 0:  \n",
    "            \n",
    "#             with open(f'{minout}/Empty_files.txt',\"w\") as outfile:\n",
    "#                 for empty in empties:\n",
    "#                     outfile.write(os.path.basename(empty)+\"\\n\")            \n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:45.342510Z",
     "start_time": "2020-09-15T20:32:45.203717Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/Z_OUT_All_Bacteria_bin_species\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank species: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/Z_OUT_All_Bacteria_bin_species\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank species: Length: 7\n",
      "Running Ttest with 7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/m/anaconda3/lib/python3.7/site-packages/scipy/stats/morestats.py:2352: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  W = numer / denom\n",
      "/Users/m/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:205: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_with_indexer(indexer, value)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "NF_path = \"/Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master/\"\n",
    "NF_out = f\"{NF_path}/NF_OUT\" \n",
    "col_data = f\"{NF_path}/sample_table.txt\"\n",
    "\n",
    "\n",
    "df_col = pd.read_csv(col_data,sep=\"\\t\",names=[\"sample\",\"condition\"])\n",
    "items = list(df_col[\"condition\"].unique())\n",
    "conditions = [(items[i],items[j]) for i in range(len(items)) for j in range(i+1, len(items))]\n",
    "\n",
    "\n",
    "permutation = conditions[0]\n",
    "control = permutation[0]\n",
    "treatment = permutation[1]\n",
    "\n",
    "\n",
    "\n",
    "# searchlist = [(True,\"phylum\",\"NA\")]\n",
    "# searchlist = [(True,\"superkingdom\",\"NA\")] # Everything without a superkingdom\n",
    "# searchlist = [(True,\"superkingdom\",\"Bacteria\")] # Everything without a superkingdom\n",
    "\n",
    "\n",
    "\n",
    "control = \"one_POLYA\"\n",
    "treatment = \"TW4_POLYA\"\n",
    "\n",
    "\n",
    "searchlist = None\n",
    "\n",
    "\n",
    "searchlist = [(True,\"superkingdom\",\"Bacteria\")]\n",
    "\n",
    "binRank = \"species\"\n",
    "\n",
    "title = \"Z_OUT_All_Bacteria_bin_species\"\n",
    "\n",
    "T = FilterTtest(NF_out=NF_out,\n",
    "                NF_path = NF_path,\n",
    "             col_data = col_data, \n",
    "             title=title, \n",
    "             searchlist=searchlist,\n",
    "             control = control, \n",
    "             treatment = treatment, \n",
    "             binRank = binRank,\n",
    "             min_length = 100,\n",
    "             p_adj = 0.05\n",
    "            )\n",
    "\n",
    "T.runSearch()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:45.844094Z",
     "start_time": "2020-09-15T20:32:45.827402Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>toy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alfred</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Batman</td>\n",
       "      <td>Batmobile</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name        toy\n",
       "0  Alfred        NaN\n",
       "1  Batman  Batmobile"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame({\"name\": ['Alfred', 'Batman', np.nan],\n",
    "\n",
    "                   \"toy\": [np.nan, 'Batmobile', 'Bullwhip']})\n",
    "\n",
    "\n",
    "df\n",
    "\n",
    "df_nan = df[~df[\"name\"].isnull()]\n",
    "df_nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:46.819533Z",
     "start_time": "2020-09-15T20:32:46.816437Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# # Ranks\n",
    "ranks = ['superkingdom',\n",
    " 'kingdom',\n",
    " 'phylum',\n",
    " 'order',\n",
    " 'family',\n",
    " 'genus',\n",
    " 'species',\"name\"]\n",
    "\n",
    "\n",
    "superkingdoms = ['Bacteria', 'Viruses', 'Eukaryota', \"NA\", 'Archaea']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:50.430665Z",
     "start_time": "2020-09-15T20:32:47.390415Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**** working on Bacteria\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_superkingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank superkingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_superkingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank superkingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_kingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank kingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_kingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank kingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_phylum\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank phylum: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_phylum\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank phylum: Length: 3\n",
      "Running Ttest with 3\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_order\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank order: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_order\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank order: Length: 4\n",
      "Running Ttest with 4\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_family\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank family: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_family\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank family: Length: 5\n",
      "Running Ttest with 5\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_genus\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank genus: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_genus\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank genus: Length: 5\n",
      "Running Ttest with 5\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_species\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank species: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_species\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank species: Length: 7\n",
      "Running Ttest with 7\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_name\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank name: Length: 2\n",
      "Running Ttest with 2\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Bacteria_Bin_name\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Bacteria\n",
      "Binning by Rank name: Length: 14\n",
      "Running Ttest with 14\n",
      "**** working on Viruses\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_superkingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_superkingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank superkingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_kingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_kingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank kingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_phylum\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_phylum\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank phylum: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_order\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_order\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank order: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_family\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_family\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank family: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_genus\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_genus\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank genus: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_species\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_species\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank species: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_name\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Viruses\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Viruses_Bin_name\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Viruses\n",
      "Binning by Rank name: Length: 1\n",
      "Running Ttest with 1\n",
      "**** working on Eukaryota\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_superkingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_superkingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank superkingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_kingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_kingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank kingdom: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_phylum\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_phylum\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank phylum: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_order\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_order\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank order: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_family\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_family\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank family: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_genus\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_genus\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank genus: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_species\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_species\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank species: Length: 1\n",
      "Running Ttest with 1\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_name\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Eukaryota_Bin_name\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Eukaryota\n",
      "Binning by Rank name: Length: 1\n",
      "Running Ttest with 1\n",
      "**** working on NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_superkingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_superkingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_kingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_kingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_phylum\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_phylum\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_order\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_order\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_family\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_family\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_genus\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_genus\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_species\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_species\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_name\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:NA\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_NA_Bin_name\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:NA\n",
      "**** working on Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_superkingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_superkingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_kingdom\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_kingdom\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_phylum\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_phylum\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_order\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_order\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_family\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_family\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_genus\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_genus\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_species\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_species\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n",
      "Running Ttest with 6\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_name\n",
      " Working on file dark_all_FG_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 2\n",
      "Searching with True:superkingdom:Archaea\n",
      "---------------\n",
      "out: /Users/m/Google_Drive/Scripts/2019/biome/biome_shared/general/datasets/chlamydia/RNAseq-Biome-master//NF_OUT/FINAL_OUT/A_B_Archaea_Bin_name\n",
      " Working on file all_pileupCoverageNormalizedMatched.txt\n",
      "Starting contigs: 1199\n",
      "Searching with True:superkingdom:Archaea\n"
     ]
    }
   ],
   "source": [
    "for superkingdom in superkingdoms:\n",
    "    print(f\"**** working on {superkingdom}\")\n",
    "    searchlist = [(True,\"superkingdom\",superkingdom)]\n",
    "    for binRank in ranks:\n",
    "  \n",
    "        makeFolders(f\"OUT_{superkingdom}\")\n",
    "        title = f\"A_B_{superkingdom}_Bin_{binRank}\"\n",
    "        \n",
    "        T = FilterTtest(NF_out=NF_out,\n",
    "                        NF_path = NF_path,\n",
    "                     col_data = col_data, \n",
    "                     title=title, \n",
    "                     searchlist=searchlist,\n",
    "                     control = control, \n",
    "                     treatment = treatment, \n",
    "                     binRank = binRank,\n",
    "                     min_length = 100,\n",
    "                     p_adj = 0.05\n",
    "                    )\n",
    "\n",
    "        T.runSearch()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:52.211954Z",
     "start_time": "2020-09-15T20:32:52.208184Z"
    }
   },
   "outputs": [],
   "source": [
    "#export\n",
    "import argparse\n",
    "def parse_arguments():\n",
    "        parser = argparse.ArgumentParser(description='filter quantify and graph')\n",
    "        parser.add_argument('--Nextflow_Out', action= 'store', metavar='--Nextflow_Out') \n",
    "        parser.add_argument('--Nextflow_path', action= 'store', metavar='--Nextflow_path') \n",
    "        args = parser.parse_args()\n",
    "        return args\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:53.343061Z",
     "start_time": "2020-09-15T20:32:53.328675Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--Nextflow_Out --Nextflow_Out]\n",
      "                             [--Nextflow_path --Nextflow_path]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /Users/m/Library/Jupyter/runtime/kernel-1d46cadb-aa23-4d58-8648-0ed0bcaf6e06.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/m/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3333: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "#export\n",
    "if __name__==\"__main__\":\n",
    "    args = parse_arguments()\n",
    "\n",
    "    NF_out = args.Nextflow_Out\n",
    "    NF_path = args.Nextflow_path\n",
    "    col_data = f\"{NF_path}/sample_table.txt\"\n",
    "\n",
    "    \n",
    "    levels = [\"all\",\"dark_all_FG\"]\n",
    "\n",
    "    ranks = ['superkingdom',\n",
    " 'kingdom',\n",
    " 'phylum',\n",
    " 'order',\n",
    " 'family',\n",
    " 'genus',\n",
    " 'species',\"name\"]\n",
    "\n",
    "\n",
    "    superkingdoms = ['Bacteria', 'Viruses', 'Eukaryota', 'NA', 'Archaea']    \n",
    "    \n",
    "    \n",
    "    # Make permutation of conditions\n",
    "    col_data = f\"{NF_path}/sample_table.txt\"\n",
    "    df_col = pd.read_csv(col_data,sep=\"\\t\",names=[\"sample\",\"condition\"])\n",
    "    items = list(df_col[\"condition\"].unique())\n",
    "    conditions = [(items[i],items[j]) for i in range(len(items)) for j in range(i+1, len(items))]\n",
    "    \n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "    # This will run all contigs \n",
    "    searchlist = None\n",
    "    binRank = None\n",
    "\n",
    "    \n",
    "    \n",
    "    for permutation in conditions:\n",
    "        control = permutation[0]\n",
    "        treatment = permutation[1]\n",
    "\n",
    "        title = f\"{control}_{treatment}_OUT_ALL_CONTIGS\"\n",
    "\n",
    "        T = FilterTtest(NF_out=NF_out, \n",
    "                     NF_path=NF_path,\n",
    "                     col_data = col_data, \n",
    "                     title=title, \n",
    "                     searchlist=searchlist,\n",
    "                     control = control, \n",
    "                     treatment = treatment, \n",
    "                     binRank = binRank,\n",
    "                     min_length = 100,\n",
    "                     p_adj = 0.05\n",
    "                    )\n",
    "\n",
    "        T.runSearch()  \n",
    "        \n",
    "\n",
    "\n",
    "    for permutation in conditions:\n",
    "        control = permutation[0]\n",
    "        treatment = permutation[1]\n",
    "\n",
    "        for superkingdom in superkingdoms:\n",
    "            print(f\"**** working on {superkingdom}\")\n",
    "            searchlist = [(True,\"superkingdom\",superkingdom)]\n",
    "            for binRank in ranks:\n",
    "\n",
    "                makeFolders(f\"OUT_{superkingdom}\")\n",
    "                title = f\"{control}_{treatment}_{superkingdom}_Bin_{binRank}\"\n",
    "\n",
    "                T = FilterTtest(NF_out=NF_out, \n",
    "                             NF_path=NF_path,\n",
    "                             col_data = col_data, \n",
    "                             title=title, \n",
    "                             searchlist=searchlist,\n",
    "                             control = control, \n",
    "                             treatment = treatment, \n",
    "                             binRank = binRank,\n",
    "                             min_length = 100,\n",
    "                             p_adj = 0.05\n",
    "                            )\n",
    "\n",
    "                T.runSearch()\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-15T20:32:56.598353Z",
     "start_time": "2020-09-15T20:32:56.340804Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 5FilterTtest.ipynb to nb_5FilterTtest.py\r\n"
     ]
    }
   ],
   "source": [
    "import fire\n",
    "!python notebook2script.py 5FilterTtest.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
